core/src/test/scala/io/projectglow/gff/GffReaderSuite.scala

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright 2019 The Glow Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.projectglow.gff
+
+import io.projectglow.common.FeatureSchemas._
+import io.projectglow.sql.GlowBaseTest
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types._
+
+class GffReaderSuite extends GlowBaseTest {
+  lazy val testRoot = s"$testDataHome/gff"
+  lazy val testGff3 = s"$testRoot/test_gff_with_fasta.gff"
+  lazy val testGff3Gzip = s"$testRoot/test_gff_with_fasta_gzip.gff.gz"
+  lazy val testGff3Bgzip = s"$testRoot/test_gff_with_fasta_bgzip.gff.bgz"
+  lazy val testGff3BgzipWithGzSuffix = s"$testRoot/test_gff_with_fasta_bgzip.gff.gz"
+  lazy val testGffEmpty = s"$testRoot/test_gff_empty.gff"
+
+  private val sourceName = "gff"
+
+  private val testOfficialFields = Seq(
+    StructField("ID", StringType),
+    StructField("Name", StringType),
+    StructField("Parent", ArrayType(StringType)),
+    StructField("Dbxref", ArrayType(StringType)),
+    StructField("Is_circular", BooleanType)
+  )
+
+  private val testUnofficialFields = Seq(
+    StructField("chromosome", StringType),
+    StructField("description", StringType),
+    StructField("gbkey", StringType),
+    StructField("gene", StringType),
+    StructField("gene_biotype", StringType),
+    StructField("gene_synonym", StringType),
+    StructField("genome", StringType),
+    StructField("mol_type", StringType),
+    StructField("product", StringType),
+    StructField("pseudo", StringType),
+    StructField("transcript_id", StringType)
+  )
+
+  private val testSchemaAllSmall = StructType(
+    gffBaseSchema.fields :+
+    StructField("dbxref", ArrayType(StringType))
+    :+ StructField("iscircular", BooleanType)
+  )
+
+  private val testSchemaMixedCase = StructType(
+    gffBaseSchema.fields :+
+    StructField("Dbxref", ArrayType(StringType))
+    :+ StructField("isCircular", BooleanType)
+  )
+
+  private val testSchemaWithUnderscore = StructType(
+    gffBaseSchema.fields :+
+    StructField("dbx_Ref", ArrayType(StringType))
+    :+ StructField("is_Circular", BooleanType)
+  )
+
+  private val testSchemaExpected = StructType(
+    gffBaseSchema.fields :+
+    StructField("Dbxref", ArrayType(StringType))
+    :+ StructField("Is_circular", BooleanType)
+  )
+
+  test("Schema inference") {
```

Comment:
Are there any particular edge cases we should look for here? E.g., no attributes fields, attributes fields containing weird characters (can there be escaped `=` or `;` inside attribute values)?